module/VuFind/src/VuFind/Db/Row/Resource.php

```diff
@@ -161,6 +161,46 @@ public function addComment($comment, $user)
         return $row->id;
     }
 
+    /**
+     * Add or update user's rating for the current resource.
+     *
+     * @param int  $userId User ID
+     * @param ?int $rating Rating
+     *
+     * @throws LoginRequiredException
+     * @throws \Exception
+     * @return int ID of rating added, deleted or updated
+     */
+    public function addOrUpdateRating(int $userId, ?int $rating): int
+    {
+        if (null !== $rating && ($rating < 0 || $rating > 100)) {
+            throw new \Exception('Rating value out of range');
+        }
+
+        $ratings = $this->getDbTable('Ratings');
+        $callback = function ($select) use ($userId) {
+            $select->where->equalTo('ratings.resource_id', $this->id);
+            $select->where->equalTo('ratings.user_id', $userId);
+        };
+        if ($existing = $ratings->select($callback)->current()) {
+            if (null === $rating) {
+                $existing->delete();
+            } else {
+                $existing->rating = $rating;
+                $existing->save();
+            }
+            return $existing->id;
+        }
+
+        $row = $ratings->createRow();
```

Comment:
Should we add a safety to not create a row if `$rating` is null? Should we return null in that case?